helix-event/src/registry.rs

```diff
@@ -0,0 +1,127 @@
+//! A global registry where events are registered and can be
+//! subsribed to by registering hooks. The registry identifies event
+//! types using their type name so multiple event with the same type name
+//! may not be registered (will cause a panic to ensure soudness)
+
+use std::any::TypeId;
+
+use anyhow::{bail, Result};
+use hashbrown::hash_map::Entry;
+use hashbrown::HashMap;
+use parking_lot::RwLock;
+
+use crate::hook::ErasedHook;
+use crate::runtime_local;
+
+pub struct Registry {
+    events: HashMap<&'static str, TypeId, ahash::RandomState>,
+    handlers: HashMap<&'static str, Vec<ErasedHook>, ahash::RandomState>,
+}
+
+impl Registry {
+    pub fn register_event<E: Event + 'static>(&mut self) {
+        let ty = TypeId::of::<E>();
+        assert_eq!(ty, TypeId::of::<E::Static>());
+        match self.events.entry(E::ID) {
+            Entry::Occupied(entry) => {
+                if entry.get() == &ty {
+                    // don't warn during tests to avoid log spam
+                    #[cfg(not(feature = "integration_test"))]
+                    log::warn!("Event {} was registered multiple times", E::ID);
```

Comment:
Is there some reason we want to allow this with a warning instead of just panicking? My first thought is that registering the same event type multiple times is probably just a programer mistake, but maybe I'm not thinking of something.

Reply:
yeah I changed this to panic. I originally had the warning here for integration tests (quite a few of them start an application multiple times) but I added the cfg later to avoid log spam anyway so now it might aswell be a panic